File: lib/features/p115/data/p115_client.dart
import 'dart:convert';

import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../../core/diagnostics/diagnostic_log.dart';
import '../../../core/net/media_proxy.dart';
import '../../../core/scanner/file_classifier.dart';
import '../../browse/data/remote_models.dart';
import 'p115_cipher.dart';
import 'p115_cookie_store.dart';

class P115Exception implements Exception {
  const P115Exception(this.message);
  final String message;
  @override
  String toString() => message;
}

class P115AuthExpiredException extends P115Exception {
  const P115AuthExpiredException() : super('115 登录已失效');
}

/// 115 returned an HTML page (verification / rate-limit / WAF) instead of JSON.
/// Distinct from auth expiry: the cookie is still valid, so callers must NOT
/// clear it — surface a retry hint instead.
class P115BlockedException extends P115Exception {
  const P115BlockedException()
    : super('115 暂时限制了访问，可能是短时间请求过多或需在网页端验证，请稍后重试或切换网络。');
}

class P115Client {
  P115Client({
    required this.cookieStore,
    Dio? dio,
    this.apiMinInterval = const Duration(milliseconds: 1200),
  }) : _dio =
           dio ??
           Dio(
             BaseOptions(
               connectTimeout: const Duration(seconds: 10),
               receiveTimeout: const Duration(seconds: 25),
             ),
           );

  final P115CookieStore cookieStore;
  final Dio _dio;
  final Duration apiMinInterval;

  Future<void> _apiGate = Future<void>.value();
  DateTime? _lastApiAt;
  var _resolveSeq = 0;

  static const sourceId = 'p115';
  static const sourceName = '115 网盘';
  static const _downloadUserAgent = 'Mozilla/5.0 115Browser/30.4.0';

  Future<List<RemoteEntry>> list(String cid) async {
    final out = <RemoteEntry>[];
    var offset = 0;
    const limit = 500;
    while (true) {
      final json = await _getJson(
        'https://webapi.115.com/files',
        query: {
          'aid': 1,
          'cid': cid,
          'offset': offset,
          'limit': limit,
          'show_dir': 1,
          'fc_mix': 0,
          'natsort': 1,
          'format': 'json',
        },
      );
      final page = mapEntries(json);
      out.addAll(page);
      final count = int.tryParse('${json['count'] ?? json['total'] ?? 0}') ?? 0;
      offset += page.length;
      if (page.length < limit || (count > 0 && offset >= count)) break;
    }
    return out;
  }

  Future<ResolvedMediaUrl> resolveVideoUrl(String pickcode) async {
    final direct = await _resolveDirect(pickcode);
    // fvp/FFmpeg won't forward a Cookie header at all, so stream playback
    // through the local proxy which injects the auth headers per request.
    final proxied = await MediaProxy.instance.wrap(direct.url, direct.headers);
    return ResolvedMediaUrl(url: proxied.url, release: proxied.release);
  }

  Future<ResolvedMediaUrl> resolveAudioUrl(String pickcode) async {
    final direct = await _resolveDirect(pickcode);
    final proxied = await MediaProxy.instance.wrapAudio(
      direct.url,
      direct.headers,
    );
    return ResolvedMediaUrl(url: proxied.url, release: proxied.release);
  }

  /// Resolves the signed CDN direct link plus the headers it needs. The CDN
  /// 403s "no cookie value" unless we send the session cookie PLUS the signed
  /// anti-leech cookies 115 sets during the download redirect (acw_tc + a
  /// dynamic one), with a 115 referer.
  Future<({Uri url, Map<String, String> headers})> _resolveDirect(
    String pickcode,
  ) async {
    final resolveId = ++_resolveSeq;
    DiagnosticLog.write('p115', 'resolve_start', {
      'resolveId': resolveId,
      'pickcodeTail': _tail(pickcode),
    });
    try {
      final resolved = await _throttledApi(() async {
        final cookie = await _cookie();
        final encrypted = P115Cipher.encryptJson({'pickcode': pickcode});
        final res = await _dio.post<dynamic>(
          'https://proapi.115.com/app/chrome/downurl',
          data: {'data': encrypted},
          options: Options(
            contentType: Headers.formUrlEncodedContentType,
            headers: {
              'Cookie': cookie.header,
              'User-Agent': _downloadUserAgent,
            },
            followRedirects: false,
            validateStatus: (s) => s != null && s < 500,
          ),
        );
        DiagnosticLog.write('p115', 'downurl_response', {
          'resolveId': resolveId,
          'status': res.statusCode,
          'setCookieNames': _cookieNamesFromSetCookies(
            res.headers['set-cookie'] ?? const [],
          ),
          'hasLocation': res.headers.value('location') != null,
        });
        if (res.statusCode == 401 || res.statusCode == 403) {
          throw const P115AuthExpiredException();
        }
        final followed = await _followDownurl(res, cookie);
        return (cookie: cookie, followed: followed);
      });
      final followed = resolved.followed;
      final cookie = resolved.cookie;
      final json = _asJson(followed.body);
      if (!_truthy(json['state'])) {
        if (_authExpired(json)) throw const P115AuthExpiredException();
        throw P115Exception('${json['error'] ?? json['message'] ?? '获取直链失败'}');
      }
      final data = jsonDecode(P115Cipher.decryptToString('${json['data']}'));
      final url = _extractDownloadUrl(data);
      final uri = Uri.parse(url);
      final proxyCookie = _mergeCookieHeader(cookie.header, followed.cookies);
      DiagnosticLog.write('p115', 'resolve_done', {
        'resolveId': resolveId,
        'hopCount': followed.hopCount,
        'lastStatus': followed.lastStatus,
        'downloadScheme': uri.scheme,
        'downloadHost': uri.host,
        'downloadQueryKeys': uri.queryParameters.keys.toList()..sort(),
        'setCookieNames': _cookieNamesFromSetCookies(followed.cookies),
        'proxyCookieNames': _cookieNamesFromHeader(proxyCookie),
      });
      return (
        url: uri,
        headers: {
          'User-Agent': _downloadUserAgent,
          'Cookie': proxyCookie,
          'Referer': 'https://115.com/',
        },
      );
    } catch (e) {
      DiagnosticLog.write('p115', 'resolve_error', {
        'resolveId': resolveId,
        'errorType': '${e.runtimeType}',
        'message': '$e',
      });
      if (e is DioException) throw _networkError(e);
      rethrow;
    }
  }

  /// One-shot whole-file download (subtitles). Goes straight to the CDN with the
  /// auth headers — dio *can* send a Cookie, so no proxy. Routing this through
  /// [MediaProxy] would bound the response to one 4 MiB block and answer 206,
  /// which this 200-expecting path treats as a failure.
  Future<List<int>> getBytesByPickcode(String pickcode) async {
    final direct = await _resolveDirect(pickcode);
    final res = await _dio.get<List<int>>(
      direct.url.toString(),
      options: Options(
        responseType: ResponseType.bytes,
        headers: direct.headers,
        validateStatus: (s) => s != null && s < 500,
      ),
    );
    if (res.statusCode == 401 || res.statusCode == 403) {
      throw const P115AuthExpiredException();
    }
    if (res.statusCode != 200 && res.statusCode != 206) {
      throw P115Exception('115 文件下载失败：${res.statusCode}');
    }
    return res.data!;
  }

  /// 115 proapi 302-redirects the downurl POST to a `dl302` gateway that serves
  /// the encrypted JSON. dart:io doesn't auto-follow a POST redirect, so chase
  /// the `Location` chain manually (GET) and return the JSON-bearing body.
  Future<_DownurlResult> _followDownurl(
    Response<dynamic> res,
    P115Cookie cookie,
  ) async {
    var current = res;
    var hopCount = 0;
    final setCookies = <String>[];
    setCookies.addAll(current.headers['set-cookie'] ?? const []);
    for (var hop = 0; hop < 5; hop++) {
      final location = current.headers.value('location');
      if (location == null || location.isEmpty) {
        return _DownurlResult(
          body: current.data,
          cookies: setCookies,
          hopCount: hopCount,
          lastStatus: current.statusCode,
        );
      }
      hopCount++;
      current = await _dio.get<dynamic>(
        location,
        options: Options(
          headers: {
            'Cookie': _mergeCookieHeader(cookie.header, setCookies),
            'User-Agent': _downloadUserAgent,
          },
          followRedirects: false,
          responseType: ResponseType.plain,
          validateStatus: (s) => s != null && s < 500,
        ),
      );
      setCookies.addAll(current.headers['set-cookie'] ?? const []);
      if (current.statusCode == 401 || current.statusCode == 403) {
        throw const P115AuthExpiredException();
      }
    }
    return _DownurlResult(
      body: current.data,
      cookies: setCookies,
      hopCount: hopCount,
      lastStatus: current.statusCode,
    );
  }

  static List<RemoteEntry> mapEntries(Map<String, dynamic> json) {
    final rows = (json['data'] as List? ?? const []).cast<dynamic>();
    final entries = rows.map((raw) {
      final item = Map<String, dynamic>.from(raw as Map);
      final fid = '${item['fid'] ?? ''}';
      final isFile = fid.isNotEmpty && fid != '0' && fid != 'null';
      final name = '${item['n'] ?? item['name'] ?? item['file_name']}';
      if (!isFile) {
        final cid = '${item['cid'] ?? item['file_id']}';
        return RemoteEntry(
          id: cid,
          path: cid,
          name: name,
          kind: RemoteEntryKind.folder,
          sourceId: sourceId,
        );
      }
      return RemoteEntry(
        id: fid,
        path: fid,
        name: name,
        kind: remoteEntryKindFromFileKind(FileClassifier.classify(name)),
        size: int.tryParse('${item['s'] ?? item['file_size'] ?? ''}'),
        pickcode: '${item['pc'] ?? item['pick_code']}',
        sourceId: sourceId,
      );
    }).toList();
    entries.sort((a, b) {
      if (a.isFolder != b.isFolder) return a.isFolder ? -1 : 1;
      return a.name.toLowerCase().compareTo(b.name.toLowerCase());
    });
    return entries;
  }

  Future<Map<String, dynamic>> _getJson(
    String url, {
    required Map<String, dynamic> query,
  }) async {
    return _throttledApi(() async {
      final cookie = await _cookie();
      final Response<dynamic> res;
      try {
        res = await _dio.get<dynamic>(
          url,
          queryParameters: query,
          options: Options(
            headers: {'Cookie': cookie.header},
            validateStatus: (s) => s != null && s < 500,
          ),
        );
      } on DioException catch (e) {
        throw _networkError(e);
      }
      if (res.statusCode == 401 || res.statusCode == 403) {
        throw const P115AuthExpiredException();
      }
      final json = _asJson(res.data);
      if (!_truthy(json['state'])) {
        if (_authExpired(json)) throw const P115AuthExpiredException();
        throw P115Exception(
          '${json['error'] ?? json['message'] ?? '115 请求失败'}',
        );
      }
      return json;
    });
  }

  Future<P115Cookie> _cookie() async {
    final cookie = await cookieStore.read();
    if (cookie == null) throw const P115AuthExpiredException();
    return cookie;
  }

  // Connection-level dio failures carry no useful message for users; map them
  // to the same friendly wording WebDAV uses so offline browse/playback reads
  // as a network problem, not a stack-trace string.
  static P115Exception _networkError(DioException e) {
    switch (e.type) {
      case DioExceptionType.connectionTimeout:
      case DioExceptionType.sendTimeout:
      case DioExceptionType.receiveTimeout:
        return const P115Exception('连接 115 超时，请检查网络后重试');
      case DioExceptionType.connectionError:
        return const P115Exception('无法连接到 115，请检查网络');
      case DioExceptionType.badCertificate:
        return const P115Exception('115 证书校验失败');
      default:
        return P115Exception('115 请求失败：${e.message ?? e.type.name}');
    }
  }

  // 115 throttles bursts across cookie-backed API endpoints, not just /files.
  // Keep directory listing, import subtitle downurl resolution, browse resolve,
  // and playback resolve on one serial rhythm so separate UI flows cannot burst.
  Future<T> _throttledApi<T>(Future<T> Function() request) {
    final run = _apiGate.then((_) async {
      final last = _lastApiAt;
      if (last != null) {
        final wait = apiMinInterval - DateTime.now().difference(last);
        if (wait > Duration.zero) await Future<void>.delayed(wait);
      }
      _lastApiAt = DateTime.now();
      return request();
    });
    _apiGate = run.then<void>((_) {}, onError: (_) {});
    return run;
  }

  Map<String, dynamic> _asJson(dynamic data) {
    if (data is Map<String, dynamic>) return data;
    if (data is Map) return Map<String, dynamic>.from(data);
    try {
      return Map<String, dynamic>.from(jsonDecode('$data') as Map);
    } on FormatException {
      throw const P115BlockedException();
    }
  }

  static bool _truthy(Object? value) => value == true || value == 1;

  static bool _authExpired(Map<String, dynamic> json) {
    final code = '${json['errno'] ?? json['errNo'] ?? json['code'] ?? ''}';
    return code == '401' || code == '403' || code == '911';
  }

  static String _mergeCookieHeader(
    String baseHeader,
    Iterable<String> setCookies,
  ) {
    final values = <String, String>{};

    void putPair(String pair) {
      final trimmed = pair.trim();
      final i = trimmed.indexOf('=');
      if (i < 0) return;
      values[trimmed.substring(0, i)] = trimmed.substring(i + 1);
    }

    for (final pair in baseHeader.split(';')) {
      putPair(pair);
    }
    for (final header in setCookies) {
      putPair(header.split(';').first);
    }
    return values.entries.map((e) => '${e.key}=${e.value}').join('; ');
  }

  static List<String> _cookieNamesFromSetCookies(Iterable<String> headers) {
    final names = <String>{};
    for (final header in headers) {
      final first = header.split(';').first.trim();
      final i = first.indexOf('=');
      if (i > 0) names.add(first.substring(0, i));
    }
    return names.toList()..sort();
  }

  static List<String> _cookieNamesFromHeader(String header) {
    final names = <String>{};
    for (final pair in header.split(';')) {
      final trimmed = pair.trim();
      final i = trimmed.indexOf('=');
      if (i > 0) names.add(trimmed.substring(0, i));
    }
    return names.toList()..sort();
  }

  static String _tail(String value) {
    if (value.length <= 6) return value;
    return value.substring(value.length - 6);
  }

  static String _extractDownloadUrl(Object? data) {
    final map = Map<String, dynamic>.from(data as Map);
    final single = map['url'];
    if (single is String) return single;
    for (final value in map.values) {
      final info = Map<String, dynamic>.from(value as Map);
      final url = info['url'];
      if (url is String) return url;
      if (url is Map) return '${url['url']}';
    }
    throw const P115Exception('115 未返回可播放直链');
  }
}

class _DownurlResult {
  const _DownurlResult({
    required this.body,
    required this.cookies,
    required this.hopCount,
    required this.lastStatus,
  });

  final dynamic body;
  final List<String> cookies;
  final int hopCount;
  final int? lastStatus;
}

final p115ClientProvider = Provider<P115Client>((ref) {
  return P115Client(cookieStore: ref.watch(p115CookieStoreProvider));
});

File: native/TonariCore/Sources/TonariCore/P115/P115Client.swift
import Foundation

/// A file or folder in a remote source (115 or WebDAV).
public struct RemoteEntry: Sendable, Hashable, Identifiable, Codable {
    public enum Kind: String, Sendable, Codable {
        case folder, audio, video, image, subtitle, text, other
    }

    public let id: String
    /// 115: the folder cid or file id. WebDAV: the path.
    public let path: String
    public let name: String
    public let kind: Kind
    public let size: Int?
    /// 115 only: the handle used to fetch a file's direct link.
    public let pickcode: String?
    public let sourceId: String

    public var isFolder: Bool { kind == .folder }

    public init(id: String, path: String, name: String, kind: Kind, size: Int? = nil, pickcode: String? = nil, sourceId: String) {
        self.id = id
        self.path = path
        self.name = name
        self.kind = kind
        self.size = size
        self.pickcode = pickcode
        self.sourceId = sourceId
    }
}

/// A streamable URL plus the headers the server insists on.
public struct ResolvedMedia: Sendable {
    public let url: URL
    public let headers: [(name: String, value: String)]
    /// When a signed link stops working; nil when it doesn't expire.
    public let expiresAt: Date?
}

public enum P115Error: LocalizedError, Equatable {
    case notLoggedIn
    case authExpired
    /// 115 answered with an HTML page (verification, rate limit, WAF). The
    /// cookie is still good; the user should retry later.
    case blocked
    case failed(String)

    public var errorDescription: String? {
        switch self {
        case .notLoggedIn: "尚未登录 115"
        case .authExpired: "115 登录已失效，请重新扫码登录"
        case .blocked: "115 暂时限制了访问，可能是短时间请求过多或需在网页端验证，请稍后重试或切换网络。"
        case .failed(let message): message
        }
    }
}

/// 115 web APIs authenticated with the cookie from a TV-QR login.
public actor P115Client {
    /// One client app-wide, so every 115 request shares the same pacing.
    public static let shared = P115Client()
    public static let sourceId = "p115"
    public static let sourceName = "115 网盘"
    static let cookieKey = "p115_cookie"
    static let userAgent = "Mozilla/5.0 115Browser/30.4.0"

    private let keychain: KeychainStore
    private let session: URLSession
    private let minInterval: Duration
    private var busy = false
    private var waiting: [CheckedContinuation<Void, Never>] = []
    private var lastRequest: ContinuousClock.Instant?

    public init(keychain: KeychainStore = .shared, minInterval: Duration = .milliseconds(1200)) {
        self.keychain = keychain
        self.minInterval = minInterval
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 25
        // Cookies are assembled by hand: the CDN needs the login cookie plus
        // the anti-leech ones picked up along the downurl redirect chain.
        config.httpShouldSetCookies = false
        config.httpCookieAcceptPolicy = .never
        session = URLSession(configuration: config)
    }

    public enum LoginState: Equatable, Sendable {
        case loggedIn, loggedOut
        /// The Keychain could not be read.
        case unreadable(String)
    }

    public nonisolated var loginState: LoginState {
        do {
            return try keychain.string(for: Self.cookieKey) == nil ? .loggedOut : .loggedIn
        } catch {
            DiagnosticLog.shared.write("p115", "cookie_unreadable", ["error": "\(error)"])
            return .unreadable("\(error)")
        }
    }

    /// A folder's entries, folders first, each group by name.
    public func list(_ cid: String) async throws -> [RemoteEntry] {
        var entries: [RemoteEntry] = []
        var offset = 0
        let limit = 500
        while true {
            var components = URLComponents(string: "https://webapi.115.com/files")!
            components.queryItems = [
                "aid": "1", "cid": cid, "offset": "\(offset)", "limit": "\(limit)", "show_dir": "1",
                "fc_mix": "0", "natsort": "1", "format": "json",
            ].map { URLQueryItem(name: $0.key, value: $0.value) }
            var request = URLRequest(url: components.url!)
            request.setValue(try cookie(), forHTTPHeaderField: "Cookie")
            let json = try await throttled { try await json(request) }
            let page = Self.entries(json)
            entries += page
            offset += page.count
            let count = Self.int(json["count"] ?? json["total"]) ?? 0
            if page.count < limit || (count > 0 && offset >= count) { break }
        }
        return entries.sorted {
            $0.isFolder != $1.isFolder ? $0.isFolder : $0.name.lowercased() < $1.name.lowercased()
        }
    }

    /// The signed CDN link for a file. The CDN answers 403 "no cookie value"
    /// unless it gets the session cookie plus the anti-leech cookies 115 sets
    /// while redirecting the downurl request, and a 115 referer.
    public func resolve(pickcode: String) async throws -> ResolvedMedia {
        let sessionCookie = try cookie()
        let (json, setCookies): ([String: Any], [String]) = try await throttled {
            var request = URLRequest(url: URL(string: "https://proapi.115.com/app/chrome/downurl")!)
            request.httpMethod = "POST"
            request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
            request.setValue(sessionCookie, forHTTPHeaderField: "Cookie")
            request.setValue(Self.userAgent, forHTTPHeaderField: "User-Agent")
            let form = "data=" + P115Cipher.encrypt(["pickcode": pickcode]).addingPercentEncoding(withAllowedCharacters: .alphanumerics)!
            request.httpBody = Data(form.utf8)
            return try await followRedirects(request, sessionCookie: sessionCookie)
        }
        guard Self.truthy(json["state"]) else {
            if Self.isAuthError(json) { throw P115Error.authExpired }
            throw P115Error.failed("\(json["error"] ?? json["message"] ?? "获取直链失败")")
        }
        let decrypted = P115Cipher.decrypt(json["data"] as! String)
        let data = try JSONSerialization.jsonObject(with: Data(decrypted.utf8))
        let url = URL(string: try Self.downloadURL(data))!
        let cookie = Self.mergeCookies(sessionCookie, setCookies)
        let expiry = URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems?
            .first { $0.name == "t" }?.value.flatMap(TimeInterval.init)
        DiagnosticLog.shared.write("p115", "resolved", ["host": url.host() ?? "", "expiresIn": expiry.map { Int($0 - Date.now.timeIntervalSince1970) } ?? -1])
        return ResolvedMedia(
            url: url,
            // FFmpeg only honours a caller User-Agent after a line break, so
            // it must not come first.
            headers: [("Cookie", cookie), ("Referer", "https://115.com/"), ("User-Agent", Self.userAgent)],
            expiresAt: expiry.map(Date.init(timeIntervalSince1970:))
        )
    }

    /// Whole-file download straight from the CDN, for subtitles.
    public func download(pickcode: String) async throws -> Data {
        let media = try await resolve(pickcode: pickcode)
        var request = URLRequest(url: media.url)
        for header in media.headers { request.setValue(header.value, forHTTPHeaderField: header.name) }
        let (data, response) = try await session.data(for: request)
        let status = (response as! HTTPURLResponse).statusCode
        if status == 401 || status == 403 { throw P115Error.authExpired }
        guard status == 200 || status == 206 else { throw P115Error.failed("115 文件下载失败：\(status)") }
        return data
    }

    // MARK: - Internals

    /// One request at a time, at least `minInterval` apart: 115 throttles
    /// bursts across all cookie-backed endpoints, so listing, subtitle
    /// downloads and playback links share one rhythm.
    private func throttled<T>(_ body: () async throws -> T) async throws -> T {
        if busy {
            await withCheckedContinuation { waiting.append($0) }
        }
        busy = true
        defer {
            if waiting.isEmpty { busy = false } else { waiting.removeFirst().resume() }
        }
        if let lastRequest {
            let wait = minInterval - (ContinuousClock.now - lastRequest)
            if wait > .zero { try await Task.sleep(for: wait) }
        }
        lastRequest = .now
        return try await body()
    }

    private func cookie() throws -> String {
        guard let cookie = try keychain.string(for: Self.cookieKey) else { throw P115Error.notLoggedIn }
        return cookie
    }

    private func json(_ request: URLRequest) async throws -> [String: Any] {
        let (data, response) = try await send(request)
        try Self.checkStatus(response)
        let json = try Self.parse(data)
        guard Self.truthy(json["state"]) else {
            if Self.isAuthError(json) { throw P115Error.authExpired }
            throw P115Error.failed("\(json["error"] ?? json["message"] ?? "115 请求失败")")
        }
        return json
    }

    /// downurl answers with a redirect to a gateway serving the encrypted
    /// JSON; each hop may set cookies the CDN later requires.
    private func followRedirects(_ first: URLRequest, sessionCookie: String) async throws -> ([String: Any], [String]) {
        var (data, response) = try await send(first)
        var setCookies = Self.setCookies(response)
        for _ in 0..<5 {
            guard let location = response.value(forHTTPHeaderField: "Location"), !location.isEmpty else { break }
            var request = URLRequest(url: URL(string: location, relativeTo: response.url)!.absoluteURL)
            request.setValue(Self.mergeCookies(sessionCookie, setCookies), forHTTPHeaderField: "Cookie")
            request.setValue(Self.userAgent, forHTTPHeaderField: "User-Agent")
            (data, response) = try await send(request)
            setCookies += Self.setCookies(response)
        }
        try Self.checkStatus(response)
        return (try Self.parse(data), setCookies)
    }

    private func send(_ request: URLRequest) async throws -> (Data, HTTPURLResponse) {
        do {
            let (data, response) = try await session.data(for: request, delegate: NoRedirects())
            return (data, response as! HTTPURLResponse)
        } catch let error as URLError {
            let message = switch error.code {
            case .timedOut: "连接 115 超时，请检查网络后重试"
            case .notConnectedToInternet, .cannotConnectToHost, .networkConnectionLost, .cannotFindHost: "无法连接到 115，请检查网络"
            default: "115 请求失败：\(error.localizedDescription)"
            }
            throw P115Error.failed(message)
        }
    }

    private final class NoRedirects: NSObject, URLSessionTaskDelegate, Sendable {
        func urlSession(_ session: URLSession, task: URLSessionTask, willPerformHTTPRedirection response: HTTPURLResponse, newRequest request: URLRequest) async -> URLRequest? {
            nil
        }
    }

    private static func checkStatus(_ response: HTTPURLResponse) throws {
        if response.statusCode == 401 || response.statusCode == 403 { throw P115Error.authExpired }
    }

    private static func parse(_ data: Data) throws -> [String: Any] {
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { throw P115Error.blocked }
        return json
    }

    static func entries(_ json: [String: Any]) -> [RemoteEntry] {
        (json["data"] as? [[String: Any]] ?? []).map { item in
            let fid = item["fid"].map { "\($0)" } ?? ""
            let name = "\(item["n"] ?? item["name"] ?? item["file_name"] ?? "")"
            guard !fid.isEmpty, fid != "0" else {
                let cid = "\(item["cid"] ?? item["file_id"] ?? "")"
                return RemoteEntry(id: cid, path: cid, name: name, kind: .folder, sourceId: sourceId)
            }
            return RemoteEntry(
                id: fid, path: fid, name: name, kind: RemoteEntry.Kind(FileKind(fileName: name)),
                size: int(item["s"] ?? item["file_size"]), pickcode: "\(item["pc"] ?? item["pick_code"] ?? "")", sourceId: sourceId
            )
        }
    }

    static func downloadURL(_ data: Any) throws -> String {
        let map = data as! [String: Any]
        if let url = map["url"] as? String { return url }
        for value in map.values {
            let url = (value as? [String: Any])?["url"]
            if let url = url as? String { return url }
            if let url = (url as? [String: Any])?["url"] as? String { return url }
        }
        throw P115Error.failed("115 未返回可播放直链")
    }

    static func mergeCookies(_ base: String, _ setCookies: [String]) -> String {
        var names: [String] = []
        var values: [String: String] = [:]
        let pairs = base.split(separator: ";") + setCookies.map { $0.split(separator: ";").first ?? "" }
        for pair in pairs {
            let trimmed = pair.trimmingCharacters(in: .whitespaces)
            guard let eq = trimmed.firstIndex(of: "=") else { continue }
            let name = String(trimmed[..<eq])
            if values[name] == nil { names.append(name) }
            values[name] = String(trimmed[trimmed.index(after: eq)...])
        }
        return names.map { "\($0)=\(values[$0]!)" }.joined(separator: "; ")
    }

    private static func setCookies(_ response: HTTPURLResponse) -> [String] {
        // URLSession folds repeated Set-Cookie headers into one, comma-joined.
        HTTPCookie.cookies(withResponseHeaderFields: response.allHeaderFields as! [String: String], for: response.url!)
            .map { "\($0.name)=\($0.value)" }
    }

    static func truthy(_ value: Any?) -> Bool {
        (value as? Bool) == true || (value as? Int) == 1
    }

    private static func isAuthError(_ json: [String: Any]) -> Bool {
        ["401", "403", "911"].contains("\(json["errno"] ?? json["errNo"] ?? json["code"] ?? "")")
    }

    static func int(_ value: Any?) -> Int? {
        switch value {
        case let n as NSNumber: n.intValue
        case let s as String: Int(s)
        default: nil
        }
    }
}

extension RemoteEntry.Kind {
    init(_ kind: FileKind) {
        self = switch kind {
        case .audio: .audio
        case .video: .video
        case .image: .image
        case .subtitle: .subtitle
        case .text: .text
        case .other: .other
        }
    }
}
